apps/otlp.go

```diff
@@ -42,45 +59,54 @@ func (r ReceiverOTLP) Type() string {
 	return "otlp"
 }
 
-func (ReceiverOTLP) gmpResourceProcessors() []otel.Component {
+func (ReceiverOTLP) gmpResourceProcessors(ctx context.Context) []otel.Component {
 	// Keep in sync with logic in confgenerator/prometheus.go
-	stmt := func(target, source string) string {
-		// if cloud.provider == "gcp" && cloud.platform == "gcp_compute_engine"
-		return fmt.Sprintf(`set(%s, %s) where %s != nil and resource.attributes["cloud.platform"] == "gcp_compute_engine"`,
-			target, source, source)
-	}
-	return []otel.Component{
-		otel.GCPResourceDetector(false),
-		{
-			Type: "transform",
-			Config: map[string]interface{}{
-				"error_mode": "ignore",
-				"metric_statements": []map[string]interface{}{{
-					"context": "datapoint",
-					"statements": []string{
-						// location = cloud.availability_zone
-						stmt(`attributes["location"]`, `resource.attributes["cloud.availability_zone"]`),
-						// namespace = host.id
-						stmt(`attributes["namespace"]`, `resource.attributes["host.id"]`),
-						// cluster = "__gce__"
-						stmt(`attributes["cluster"]`, `"__gce__"`),
-						// instance_name = host.name
-						stmt(`attributes["instance_name"]`, `resource.attributes["host.name"]`),
-						// machine_type = host.type
-						stmt(`attributes["machine_type"]`, `resource.attributes["host.type"]`),
-						// job and instance should be automatically set from service labels
-					},
-				}},
+	bmsResource := platform.FromContext(ctx).ResourceOverride
```

Comment:
This is not the right way to handle this. Whether Ops Agent is running on a BMS machine is unrelated to whether the OTLP data it's processing is coming from a BMS machine.

If you want to add support for BMS->GMP resource conversion, you'll need to add the new statements *in addition* to the existing ones, not instead of them. And probably also do so unconditionally, regardless of whether Ops Agent is running on a BMS machine.

Reply:
Okay, but If the ops-agent is running on a BMS machine and its config contains OTLP receiver with metric_mode set to googlemanagedprometheus then the generated otel config will have a GCP resourcedetection processor which won't work on a BMS machine. Shouldn't I make sure that the resourcedetection processor doesn't get included into the otel config ?

Reply:
You're right that the resourcedetection processor itself needs to change. I accidentally created a new thread elaborating on that, so I'll mark this one resolved.